rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,540 @@
+with separate objects.  For runtime leading dimension, the internal
+API supports it but this will not be exposed for now. We can either
+change the API when we introduce the feature, or add a placeholder
+parameter if we are confident this feature will gain traction soon.
+
+Also, note that currently, no kernel-level cache exists for block
+level functionalities, but there are plans to add it. In the meantime,
+here are a few ways to mitigate the lack of kernel-level cache:
+- hoist block-level kernel generation for it to happen once at the
+  beggining of the application
+- to handle tails/K-blocking, if generating multiple kernels is too
+  costly, one can copy A/B blocks to a temporary buffer for
+  tails/first K-block, and rely on a single block-level kernel.
+
+
+## Transforms and transpose
+
+Transormation rountines, for example to pack data in a certain layout,
+are typically hard to implement without using intrinsics or assembly.
+To facilitate packing, we will expose an out-of-place transform
+functionality.
+
+## All-in-all
+
+### Interface proposal
+
+```c++
+
+// namespace name to be defined, leaving it general enough for additional block level APIs
+namespace dnnl {
+namespace block { 
+
+enum packing_tag {
+packed_32;
+packed_64;
+plain;
+transposed;
+}
+
+struct brgemm {
+    struct desc {
+        // Vanilla version of brgemm with no post-op or destination conversion.
+        desc(dim_t batch, dim_t M, dim_t N, dim_t K,
+                data_type dtA, dim_t ldA,
+                data_type dtB, dim_t ldB, 
+                data_type dtC, dim_t ldC,
+                float alpha, float beta);
+
+        // Advanced version with postop and datatype conversion when D type
+        // is different than C type.
+        desc(dim_t batch, dim_t M, dim_t N, dim_t K,
+                data_type dtA, dim_t ldA,
+                data_type dtB, dim_t ldB, 
+                data_type dtC, dim_t ldC,
+                data_type dtD, dim_t ldD,
+                float alpha, float beta, 
+                const brgemm_attr &attr);
+
+        // Queries for expected layouts and temporary memory
+        packing_tag get_A_tag() const;
+        packing_tag get_B_tag() const;
+        size_t get_scratchpad_size() const;
+    }
+
+    brgemm(const desc &bd); 
+    
+    // HW context handling.
+    // This currently mimics AMX (need to clarify for SME): 
+    // - Release is static
+    // - No release necessary between different calls to set_hw_context
+    void set_hw_context() const;
+    static void release_hw_context() const;
+
+    // Execution function for the vanilla brgemm variant.
+    // we take pointers to A and B as a vector of pairs, guarantees they are the same size
+    // The batch size is the size of the vector.
+    // pointers are void*, datatypes are specified in constructor
+    // Computes C = \beta C + \alpha \sum_i A_i \cdot B_i
+    void execute(const std::vector<std::pair<void *, void *>> &A_B,
```

Comment:
It's very difficult to handle STL containers inside MLIR, will you provide C interface as well? I want to check what's the C interface looks like.

Reply:
Agreed. Also, `std::vector` might incur overhead in its memory management.